text/0000-safety-tags.md

```diff
@@ -0,0 +1,481 @@
+#![clippy::safety { core::ptr::invariants::{ValidPtr, Aligned} }]
+```
+
+That's to say:
+* Tags declared or re-exported in the current module are automatically in scope: no import required.
+* Tags from other modules must be brought in with the inner-tool attribute shown above.
+* Tags are visible and available to downstream crates whenever their declaration paths are public.
+* Attempting to import a tag from a private module is a **hard error**.
+* Referencing a tag that has never been declared is also a **hard error**.
+
+[uninhabited]: https://doc.rust-lang.org/reference/glossary.html#uninhabited
+[type namespace]: https://doc.rust-lang.org/reference/names/namespaces.html
+[scopes]: https://doc.rust-lang.org/reference/names/scopes.html#item-scopes
+[visibility]: https://doc.rust-lang.org/reference/visibility-and-privacy.html
+[`UseTree`]: https://doc.rust-lang.org/reference/items/use-declarations.html
+
+Tags are treated as items so rustdoc can render their documentation and hyperlink tag references.
+And Rust-Analyzer can offer **full IDE support**: completion, go-to-definition/declaration, and
+doc-hover.
+
+Tags constitute a public API; therefore, any alteration to their declaration or definition must be
+evaluated against [Semantic Versioning][semver].
+* Adding a tag is a **minor** change.
+* Removing a tag is a **major** change.
+
+To give dependent crates time to migrate, mark obsolete tag items with `#[deprecated]`. Clippy will
+surface the deprecation warning whenever the tag is used.
+
+[semver]: https://doc.rust-lang.org/cargo/reference/semver.html
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Unstable Features
+
+Currently, safety tags requires the following unstable features
+* `#![feature(proc_macro_hygiene, stmt_expr_attributes)]` for tagging statements or expressions.
+* `#![feature(custom_inner_attributes)]` for `#![clippy::safety {}]` imports
+
+Since the safety-tag mechanism is implemented primarily in Clippy and Rust-Analyzer, no additional
+support is required from rustc.
+
+But we ask the libs team to adopt safety tags for all public `unsafe` APIs in libstd, along with
+their call sites. To enable experimentation, a nightly-only library feature
+`#![feature(safety_tags)]` should be introduced and remain unstable until the design is finalized.
+
+## Implementation in Clippy
+
+Procedure:
+
+1. Scan the crate for every item marked `#[clippy::safety::tag]`; cache the compiled tag metadata of
+   upstream dependencies under `target/` for later queries.
+2. Validate every `#![clippy::safety { ... }]` import by a reachability analysis that ensures every
+   referenced tag is defined and accessible.
+3. Verify that every unsafe call carries the required safety tags:
+   * Resolve the callee, collect its declared tags, then walk outward from the call site until the
+     function’s own signature confirms these tags are listed in its `#![clippy::safety { ... }]`
```

Comment:
On first read through, I was confused on the syntax examples separating imports (inner attribute `#![...]`) from function definitions and use sites (both outer attributes `#[...]`). 

Should this be an outer attribute `#[...]` since it's on a function signature? Or are both inner/outer attributes allowed on function definitions?

(I think this was the only place I noticed)